File: src/main/java/com/ysmull/easeshop/controller/shop/LoginController.java
package com.ysmull.easeshop.controller.shop;

import com.ysmull.easeshop.annotation.Privilege;
import com.ysmull.easeshop.model.entity.User;
import com.ysmull.easeshop.model.vo.UserVO;
import com.ysmull.easeshop.model.vo.WebResponse;
import com.ysmull.easeshop.service.UserService;
import com.ysmull.easeshop.util.UserContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * @author maoyusu
 */
@RestController
@RequestMapping("/api")
public class LoginController {


    @Autowired
    private UserService userService;

    private final Logger log = LoggerFactory.getLogger(LoginController.class);


    @PostMapping(value = "/login")
    @ResponseBody
    public WebResponse<UserVO> login(@RequestParam("username") String username,
                                     @RequestParam("password") String password,
                                     HttpServletResponse httpResponse) {
        WebResponse<UserVO> webResponse = new WebResponse<>();
        String cookieValue;
        if ((cookieValue = userService.check(username, password)) != null) {
            Cookie cookie = new Cookie("ES_token", cookieValue);
            cookie.setPath("/");
            httpResponse.addCookie(cookie);
            log.debug("add login cookie");
            UserVO userVO = new UserVO();
            userVO.setUserId(Long.parseLong(cookieValue.split(":")[0]));
            webResponse.setData(userVO);
        } else {
            webResponse.setCode(WebResponse.NEED_LOGIN);
            webResponse.setMsg("login failed!");
        }
        return webResponse;
    }

    @GetMapping(value = "/valid")
    @Privilege(login = true)
    @ResponseBody
    public WebResponse<UserVO> valid(HttpServletResponse httpResponse) {
        User user = UserContext.getCurrentUser();
        WebResponse<UserVO> webResponse = new WebResponse<>();
        UserVO userVO = new UserVO();
        userVO.setUserId(user.getId());
        userVO.setUsername(user.getUsername());
        userVO.setRole(user.getRole().ordinal());
        webResponse.setData(userVO);
        return webResponse;
    }

    @GetMapping(value = "/logout")
    public void logout(HttpServletRequest httpRequest, HttpServletResponse httpResponse) {
        Cookie[] cookies = httpRequest.getCookies();
        for (Cookie c : cookies) {
            if ("ES_token".equals(c.getName())) {
                c.setValue("");
                c.setPath("/");
                c.setMaxAge(0);
                httpResponse.addCookie(c);
                log.debug("delete login cookie");
            }
        }
    }
}

File: EaseMall-web/src/main/java/com/ysmull/easemall/controller/shop/LoginController.kt
package com.ysmull.easemall.controller.shop

import com.ysmull.easemall.annotation.Privilege
import com.ysmull.easemall.biz.UserService
import com.ysmull.easemall.model.entity.User
import com.ysmull.easemall.model.vo.UserVO
import com.ysmull.easemall.model.vo.WebResponse
import com.ysmull.easemall.util.UserContext
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*

import javax.servlet.http.Cookie
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * @author maoyusu
 */
@RestController
@RequestMapping("/api")
class LoginController(
        val userService: UserService
) {

    private val log = LoggerFactory.getLogger(LoginController::class.java)

    @PostMapping(value = ["/login"])
    @ResponseBody
    fun login(@RequestParam("username") username: String,
              @RequestParam("password") password: String,
              httpResponse: HttpServletResponse): WebResponse<UserVO> {
        val webResponse = WebResponse<UserVO>()
        val cookieValue: String? = userService.check(username, password)
        if (cookieValue != null) {
            val cookie = Cookie("ES_token", cookieValue)
            cookie.path = "/"
            httpResponse.addCookie(cookie)
            log.debug("add login cookie")
            val userVO = UserVO()
            userVO.userId = java.lang.Long.parseLong(cookieValue.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0])
            webResponse.data = userVO
        } else {
            webResponse.code = WebResponse.NO_AUTH
            webResponse.msg = "login failed!"
        }
        return webResponse
    }

    @GetMapping(value = ["/valid"])
    @Privilege(login = true)
    @ResponseBody
    fun valid(httpResponse: HttpServletResponse): WebResponse<UserVO> {
        val user = UserContext.currentUser
        val webResponse = WebResponse<UserVO>()
        val userVO = UserVO()
        userVO.userId = user.id
        userVO.username = user.username
        userVO.role = user.role?.ordinal
        webResponse.data = userVO
        return webResponse
    }

    @GetMapping(value = ["/logout"])
    fun logout(httpRequest: HttpServletRequest, httpResponse: HttpServletResponse) {
        val cookies = httpRequest.cookies
        for (c in cookies) {
            if ("ES_token" == c.name) {
                c.value = ""
                c.path = "/"
                c.maxAge = 0
                httpResponse.addCookie(c)
                log.debug("delete login cookie")
            }
        }
    }
}
